src/cli/context.js

```diff
@@ -30,120 +30,100 @@ const createMinimistOptions = require("./create-minimist-options");
-
-  context.logger = createLogger(context.argv.loglevel);
-
-  updateContextArgv(
-    context,
-    context.argv.plugin,
-    context.argv["plugin-search-dir"]
-  );
-
-  return context;
-}
-
-function initContext(context) {
-  // split into 2 step so that we could wrap this in a `try..catch` in cli/index.js
-  normalizeContextArgv(context);
-}
-
-/**
- * @param {Context} context
- * @param {string[]} plugins
- * @param {string[]=} pluginSearchDirs
- */
-function updateContextOptions(context, plugins, pluginSearchDirs) {
-  const { options: supportOptions, languages } = prettier.getSupportInfo({
-    showDeprecated: true,
-    showUnreleased: true,
-    showInternal: true,
-    plugins,
-    pluginSearchDirs,
-  });
-
-  const detailedOptionMap = normalizeDetailedOptionMap({
-    ...createDetailedOptionMap(supportOptions),
-    ...constant.options,
-  });
-
-  const detailedOptions = arrayify(detailedOptionMap, "name");
-
-  const apiDefaultOptions = {
-    ...optionsModule.hiddenDefaults,
-    ...fromPairs(
-      supportOptions
-        .filter(({ deprecated }) => !deprecated)
-        .map((option) => [option.name, option.default])
-    ),
-  };
-
-  Object.assign(context, {
-    supportOptions,
-    detailedOptions,
-    detailedOptionMap,
-    apiDefaultOptions,
-    languages,
-  });
-}
-
-/**
- * @param {Context} context
- * @param {string[]} plugins
- * @param {string[]=} pluginSearchDirs
- */
-function pushContextPlugins(context, plugins, pluginSearchDirs) {
-  context.stack.push(
-    pick(context, [
-      "supportOptions",
-      "detailedOptions",
-      "detailedOptionMap",
-      "apiDefaultOptions",
-      "languages",
-    ])
-  );
-  updateContextOptions(context, plugins, pluginSearchDirs);
-}
-
-/**
- * @param {Context} context
- */
-function popContextPlugins(context) {
-  Object.assign(context, context.stack.pop());
-}
-
-function updateContextArgv(context, plugins, pluginSearchDirs) {
-  pushContextPlugins(context, plugins, pluginSearchDirs);
-
-  const minimistOptions = createMinimistOptions(context.detailedOptions);
-  const argv = minimist(context.args, minimistOptions);
-
-  context.argv = argv;
-  context.filePatterns = argv._.map((file) => String(file));
-}
-
-function normalizeContextArgv(context, keys) {
-  const detailedOptions = !keys
-    ? context.detailedOptions
-    : context.detailedOptions.filter((option) => keys.includes(option.name));
-  const argv = !keys ? context.argv : pick(context.argv, keys);
-
-  context.argv = optionsNormalizer.normalizeCliOptions(argv, detailedOptions, {
-    logger: context.logger,
-  });
+class Context {
+  constructor(args) {
+    this.args = args;
+    this.stack = [];
+    this._updateContextArgv();
+    this._normalizeContextArgv(["loglevel", "plugin", "plugin-search-dir"]);
+    this.logger = createLogger(this.argv.loglevel);
+    this._updateContextArgv(this.argv.plugin, this.argv["plugin-search-dir"]);
+  }
+
+  initContext() {
+    this._normalizeContextArgv();
+  }
```

Comment:
As a class, this init method seems out of place. This could be done from within the constructor. Especially since the constructor is itself calling `_normalizeContextArgv` (with arguments), some consolidation could be done.

Reply:
I forgot to move a comment about `initContext` (c1350c1). We should wrap this in `try...catch`.

Reply:
But we can move `const context = new core.Context(args);` into `try...catch`, right?

Reply:
We need to run `context.logger.error(error.message);` in `catch...` so we cannot do it.

Reply:
I haven't look the code, is logger somehow related to context? Can we seperate them? Anyway maybe in another PR.